### PR TITLE
Update ProcessStats query by using API instead of parsing ps command.

### PR DESCRIPTION
**Pull request by FengPan-Frank**
Now ProcessStats query parses ps command and populate STATE_DB, "ps" command is somewhat fragile, its format might be impacted by relevant environment change. This PR uses /proc instead which is stable and verified via systemctl procdockerstatsd.service.

Before update:
Mar 18 09:12:27.864031 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data cid  '1424'
Mar 18 09:12:27.864091 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data uid '0'
Mar 18 09:12:27.864188 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data ppid '1'
Mar 18 09:12:27.864276 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data cpu '0.0'
Mar 18 09:12:27.864367 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data mem '0.0'
Mar 18 09:12:27.864456 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data stime 'Mar15'
Mar 18 09:12:27.864545 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data tty 'tty1'
Mar 18 09:12:27.864698 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data time '00:00:00'
Mar 18 09:12:27.864762 str3-msn4700-04 INFO procdockerstatsd[2663305]: process data cmd '/sbin/agetty -o -p -- \u --noclear - linux'

After update:
Mar 18 12:02:00.784073 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data cid  '1424'
Mar 18 12:02:00.784164 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data uid  '0'
Mar 18 12:02:00.784304 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data ppid  '1'
Mar 18 12:02:00.784446 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data cpu  '0.0'
Mar 18 12:02:00.784585 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data mem  '0.0'
Mar 18 12:02:00.784743 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data stime  'Mar15'
Mar 18 12:02:00.784883 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data tty  '/dev/tty1'
Mar 18 12:02:00.785023 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data time  '00:00:00'
Mar 18 12:02:00.785164 str3-msn4700-04 INFO procdockerstatsd[2791034]: process data cmd  '/sbin/agetty -o -p -- \u --noclear - linux'

### CODE REVIEW

File: scripts/procdockerstatsd [hunk 1]
```diff
@@ -136,18 +137,30 @@ class ProcDockerStats(daemon_base.DaemonBase):
...
-        if any(exitcode):
-            cmd = ' | '.join([' '.join(cmd0), ' '.join(cmd1)])
-            self.log_error("Error running command '{}'".format(cmd))
-            data = None
-        processdata = self.format_process_cmd_output(data)
-        value = ""
+        processdata = []
+        for process in psutil.process_iter(['pid', 'ppid', 'memory_percent', 'cpu_percent', 'create_time', 'cmdline']):
+            try:
+                uid = process.uids()[0]
```
Comment (qiluo-msft) [inline scripts/procdockerstatsd:143]: any chance of index out of range?

Comment (FengPan-Frank) [reply scripts/procdockerstatsd:143]: uids() will return the user ids associated with the process. if process does not have a valid id, or accessible issue due to privileges. psutil will raise an exception which is supposed to be caught.

File: scripts/procdockerstatsd [hunk 2]
```diff
@@ -136,18 +137,30 @@ class ProcDockerStats(daemon_base.DaemonBase):
...
+                cpu = process.cpu_percent()
+                stime = process.create_time()
+                stime_formatted = datetime.fromtimestamp(stime).strftime("%b%d")
+                tty = process.terminal()
+                ttime = process.cpu_times()
+                time_formatted = str(timedelta(seconds=int(ttime.user + ttime.system)))
+                cmd = ' '.join(process.cmdline())
+
+                row = {'PID': pid, 'UID': uid, 'PPID': ppid, '%CPU': cpu, '%MEM': mem, 'STIME': stime_formatted,  'TT': tty, 'TIME': time_formatted, 'CMD': cmd}
+                processdata.append(row)
```
Comment (qiluo-msft) [inline scripts/procdockerstatsd:156]: >processdata [](http://example.com/codeflow?start=16&length=11)

processdata is only used internally. Not need to serialize or unserialize. Could you just use a python dict?

File: scripts/procdockerstatsd [hunk 3]
```diff
@@ -136,18 +137,30 @@ class ProcDockerStats(daemon_base.DaemonBase):
...
-            data = None
-        processdata = self.format_process_cmd_output(data)
-        value = ""
+        processdata = []
+        for process in psutil.process_iter(['pid', 'ppid', 'memory_percent', 'cpu_percent', 'create_time', 'cmdline']):
+            try:
+                uid = process.uids()[0]
+                pid = process.pid
+                ppid = process.ppid()
+                mem = round(process.memory_percent(), 1)
```
Comment (qiluo-msft) [inline scripts/procdockerstatsd:146]: >round [](http://example.com/codeflow?start=22&length=5)

Move data processing to below logic before writing to StateDB.

File: scripts/procdockerstatsd (hunk 2 above)
Comment (FengPan-Frank) [reply scripts/procdockerstatsd:156]: processdata is list of dict which has no serialize now.

File: scripts/procdockerstatsd (hunk 3 above)
Comment (FengPan-Frank) [reply scripts/procdockerstatsd]: done

**Review by qiluo-msft — APPROVED** [review]

File: azure-pipelines.yml [hunk 4]
```diff
@@ -64,6 +64,7 @@ stages:
     - script: |
         set -xe
         sudo pip3 install enum34
+        sudo pip install psutil
```
Comment (liushilongbuaa) [inline azure-pipelines.yml:67]: py2 is deprecated.
pip3 install

Comment (FengPan-Frank) [reply azure-pipelines.yml:67]: Thanks, fixed this on https://github.com/sonic-net/sonic-host-services/pull/108

**Pull request merged**